lib/core/src/dirty.rs

```diff
@@ -1,110 +1,539 @@
+/// to expose the same API but without requiring `self` reference to be mutable.
+#[derive(Derivative,Shrinkwrap)]
+#[derivative(Debug(bound = "T:Debug"))]
+#[derivative(Clone(bound = ""))]
+pub struct SharedDirtyFlag<T,OnSet> {
+    rc: Rc<RefCell<DirtyFlag<T,OnSet>>>
+}
+
+
+// === API ===
+
+impl<T:Default,OnSet>
+SharedDirtyFlag<T,OnSet> {
+    pub fn new(logger: Logger, on_set: OnSet) -> Self {
+        let callback = Callback(on_set);
+        let rc       = Rc::new(RefCell::new(DirtyFlag::new(logger,callback)));
+        Self { rc }
+    }
+}
+
+impl<T,OnSet>
+SharedDirtyFlag<T,OnSet> {
+    pub fn set_callback(&self, on_set:OnSet) {
+        self.rc.borrow_mut().on_set = Callback(on_set);
+    }
+}
+
+impl<T,OnSet>
+From<Rc<RefCell<DirtyFlag<T,OnSet>>>> for SharedDirtyFlag<T,OnSet> {
+    fn from(rc: Rc<RefCell<DirtyFlag<T,OnSet>>>) -> Self {
+        Self {rc}
+    }
+}
+
+
+// === Arg ===
+
+impl<T:HasArg,OnSet> HasArg for SharedDirtyFlag<T,OnSet> {
+    type Arg = Arg<T>;
+}
+
+
+// === Global Operations ===
+
+impl<T:HasUnsetAll,OnSet>
+SharedHasUnsetAll for SharedDirtyFlag<T,OnSet> {
+    fn unset_all(&self) {
+        self.rc.borrow_mut().unset_all()
+    }
+}
+
+impl<T:HasCheckAll,OnSet>
+HasCheckAll for SharedDirtyFlag<T,OnSet> {
+    fn check_all(&self) -> bool {
+        self.rc.borrow().check_all()
+    }
+}
+
+// === Check ===
+
+impl<T:DirtyFlagOps0,OnSet:Callback0>
+HasCheck0 for SharedDirtyFlag<T,OnSet> {
+    fn check (&self) -> bool { self.borrow().check()   }
+}
+
+impl<T:DirtyFlagOps1,OnSet:Callback0>
+HasCheck1 for SharedDirtyFlag<T,OnSet> {
+    fn check (&self, arg:&Arg<T>) -> bool { self.borrow().check(arg)   }
 }
 
+// === Set ===
+
+impl<T:DirtyFlagOps0,OnSet:Callback0>
+SharedHasSet0 for SharedDirtyFlag<T,OnSet> {
+    fn set (&self) { self.borrow_mut().set() }
+}
+
+impl<T:DirtyFlagOps1,OnSet:Callback0>
+SharedHasSet1 for SharedDirtyFlag<T,OnSet> {
+    fn set (&self, arg: Arg<T>) { self.borrow_mut().set(arg) }
+}
+
+// === Unset ===
+
+impl<T:HasUnset0,OnSet>
+SharedHasUnset0 for SharedDirtyFlag<T,OnSet> {
+    fn unset(&self) {
+        self.borrow_mut().unset()
+    }
+}
+
+impl<T:HasUnset1,OnSet>
+SharedHasUnset1 for SharedDirtyFlag<T,OnSet> where Arg<T>:Display {
+    fn unset(&self, arg:&Self::Arg) {
+        self.borrow_mut().unset(arg)
+    }
+}
+
+
+// === Iterators ===
+
+// FIXME: This is very error prone. Fix it after this gets resolved:
+// https://github.com/rust-lang/rust/issues/66505
+
+// [1] Please refer to `Prelude::drop_lifetime` docs to learn why it is safe to
+// use it here.
+impl<T, OnSet> SharedDirtyFlag<T,OnSet>
+where for<'t> &'t T: IntoIterator {
+    pub fn iter(&self) -> SharedDirtyFlagIter<T, OnSet> {
+        let _borrow   = self.rc.borrow();
+        let reference = unsafe { drop_lifetime(&_borrow) }; // [1]
+        let iter      = reference.data.into_iter();
+        SharedDirtyFlagIter { iter, _borrow }
+    }
```

Comment:
Very good point. I was thinking about it already several times. There are few possible solutions here:
1. This one.
2. Exposing the ref to inner data using some getter with a name telling this is a borrow.
3. Adding a `borrow` method to dirty flags and requiring its usage on call side.
4. Cloning the data in accessor (of course bad).

I suspect you tend to option 3?

Reply:
Yes, a 3. looks slightly better, because we see in the usage place that something may happen.

But still I'm afraid of some complex situations, where we're iterating over flags calling some function on each iteration, which calls another functions, which calls another, which fires some callback etc. and there we dereference another copy of SharedDirtyFlag and KABOOM! 

Actually, this kind of issues is not _hard to track down_, because the stacktrace should tell everything. But we must ensure, that each code using that is covers with automatic tests, covering each fantastic corner case.

_Or_

I'm not sure, but I think that usually when we iterate over flag content we usually also clear this flag. So maybe not _cloning_ but _taking_ shared data is an option?

Reply:
There are two things here:

1. We are using this functionality currently in 1 place in code.
2. The idea with taking ownership is really interesting one! I'll check it out.

Reply:
So, to explain this: It would be analogous to `drop` method of various containers in rust standard library. The dirty flags implementing `Iterator` could have `drop_all` which in essence calls `std::mem::take` on owned container and returns it as iterable. Then `SharedDirtyFlag` just calls this drop on borrow_mut without need of dropping lifetimes.